src/rgw/rgw_op.cc

```diff
@@ -5276,156 +5276,184 @@ void RGWCompleteMultipart::execute()
   meta_obj.set_in_extra_data(true);
   meta_obj.index_hash_source = s->object.name;
 
-  op_ret = get_obj_attrs(store, s, meta_obj, attrs);
-
-  if (op_ret < 0) {
-    ldout(s->cct, 0) << "ERROR: failed to get obj attrs, obj=" << meta_obj
-		     << " ret=" << op_ret << dendl;
-    return;
-  }
-
+  rgw_pool meta_pool;
+  rgw_raw_obj raw_obj;
+  librados::IoCtx ioctx;
+  rados::cls::lock::Lock l("RGWCompleteMultipart");
+  int max_lock_secs_mp = s->cct->_conf->rgw_mp_lock_max_time;
+
+  store->obj_to_raw((s->bucket_info).placement_rule, meta_obj, &raw_obj);
+  store->get_obj_data_pool((s->bucket_info).placement_rule,meta_obj,&meta_pool);
+  store->open_pool_ctx(meta_pool, ioctx);
+  utime_t time(max_lock_secs_mp, 0);
+  l.set_duration(time);
+  const string raw_meta_oid = raw_obj.oid;
   do {
-    op_ret = list_multipart_parts(store, s, upload_id, meta_oid, max_parts,
-				  marker, obj_parts, &marker, &truncated);
-    if (op_ret == -ENOENT) {
-      op_ret = -ERR_NO_SUCH_UPLOAD;
-    }
-    if (op_ret < 0)
-      return;
+    op_ret = l.lock_exclusive(&ioctx,raw_meta_oid);
```

Comment:
since this cls_lock call is a rados write operation, the osd will create the object if it didn't already exist - that would likely confuse the logic below. we'll need to use the `lock_exclusive(librados::ObjectWriteOperation*)` overload of this function, along with `ObjectWriteOperation::assert_exists()` to get `-ENOENT` in that case instead:
```c++
    librados::ObjectWriteOperation op;
    op.assert_exists();
    l.lock_exclusive(&op);
    op_ret = ioctx.operate(raw_meta_oid, &op);
```